adam/learner/affordances.py

```diff
@@ -0,0 +1,195 @@
+                )
+                digraph.add_node(affordance_property)
+                # This edge might not handle temporal encoding correctly?
+                digraph.add_edge(
+                    list(node.slot_fillings.values())[0],
+                    affordance_property,
+                    label=HAS_PROPERTY_LABEL,
+                )
+        return (
+            PerceptionGraph(
+                graph=digraph, dynamic=perception_graph_after_matching.dynamic
+            ),
+            immutable_new_nodes,
+        )
+
+
+@attrs(slots=True)
+class SubsetAffordanceLearner(
+    AbstractAffordanceTemplateLearner, AbstractTemplateSubsetLearner
+):
+    """
+    An implementation of `TopLevelLanguageLearner` for subset learning based approach for affordance learning.
+    """
+
+    def _update_hypothesis(
+        self,
+        previous_pattern_hypothesis: PerceptionGraphTemplate,
+        current_pattern_hypothesis: PerceptionGraphTemplate,
+    ) -> Optional[PerceptionGraphTemplate]:
+        return previous_pattern_hypothesis.intersection(
+            current_pattern_hypothesis,
+            ontology=self._ontology,
+            match_mode=MatchMode.NON_OBJECT,
+            allowed_matches=immutablesetmultidict(
+                [
+                    (node2, node1)
+                    for previous_slot, node1 in previous_pattern_hypothesis.template_variable_to_pattern_node.items()
+                    for new_slot, node2 in current_pattern_hypothesis.template_variable_to_pattern_node.items()
+                    if previous_slot == new_slot
+                ]
+            ),
+        )
+
+    def _new_concept(self, debug_string: str) -> Concept:
+        return AffordanceConcept(debug_string)
+
+    def _keep_hypothesis(
+        self,
+        *,
+        hypothesis: PerceptionGraphTemplate,
+        bound_surface_template: SurfaceTemplateBoundToSemanticNodes,
+    ) -> bool:
+        if len(bound_surface_template.slot_to_semantic_node) > 1:
+            return False
+        return len(hypothesis.graph_pattern) >= 2
+
+    def _hypotheses_from_perception(
+        self,
+        learning_state: LanguagePerceptionSemanticAlignment,
+        bound_surface_template: SurfaceTemplateBoundToSemanticNodes,
+    ) -> AbstractSet[PerceptionGraphTemplate]:
+        if len(bound_surface_template.slot_to_semantic_node) > 1:
+            raise ValueError(
+                f"Affordance learner can not extract features when more than one object node is a "
+                f"slot filler. Offending template: {bound_surface_template}"
+            )
+        return immutableset(
+            PerceptionGraphTemplate.from_graph(
+                perception_graph=candidate_attribute_meaning,
+                template_variable_to_matched_object_node=bound_surface_template.slot_to_semantic_node,
+            )
+            for candidate_attribute_meaning in _extract_candidate_attributes(
```

Comment:
Does this work? It looks like `_extract_candidate_attributes()` looks at properties, but it looks for them by checking the label against `HAS_PROPERTY_LABEL` directly. In a dynamic graph the edge label should be wrapped in a scope, which might break this?

Reply:
Actually, that function also limits itself to single-property hypotheses, which might cause problems for learning affordances. I expect affordances may need to include multiple relationships which may or may not be properties.

Reply:
Ah you make a fair point, I was trying to avoid rewriting the logic to stay internal to an object's influence on the perception graph but I _do_ want multiple properties returned as well as the stroke information. I'll just implement an extraction of affordance attributes borrowing from the attribute logic to stay internal to the object.